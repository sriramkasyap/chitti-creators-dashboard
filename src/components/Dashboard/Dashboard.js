import { useEffect } from "react";
import PropTypes from "prop-types";
import { Flex, Heading, Image } from "@chakra-ui/react";

import Card from "./Card/Card";
import NewslettersList from "../NewslettersList/NewslettersList";

import { showNotification } from "../../utils";

const Dashboard = ({ isLoading, error, cards }) => {
  useEffect(() => {
    if (error) {
      showNotification(error);
    }
  }, [error]);

  return (
    <Flex flexDir="column" w="100%" alignItems={["unset", "center"]}>
      {/* Dashboard Card Start */}
      <Flex
        flexDir={["column", "row"]}
        justifyContent={["unset", "unset", "space-between"]}
        flexWrap="wrap"
        alignItems="center"
        w="100%"
      >
        {isLoading ? (
          <Image data-testid="loading_image" src="loader_black.gif" h="5rem" />
        ) : (
          <>
            {cards?.map((card) => (
              <Card key={card.id} card={card} isLoading={isLoading} />
            ))}
          </>
        )}
      </Flex>
      {/* Dashboard Card End */}

      {/* Newsletters List Start */}
      <Flex flexDir="column" mt={10} w="100%">
        <Flex>
          <Heading mt={10} mb={5}>
            Newsletters List
          </Heading>
        </Flex>
        <Flex w="100%" overflow="auto">
          <NewslettersList />
        </Flex>
      </Flex>
      {/* Newsletters List End */}
    </Flex>
  );
};

Dashboard.propTypes = {
  isLoading: PropTypes.bool,
  error: PropTypes.string,
  cards: PropTypes.instanceOf(Array),
};

Dashboard.defaultProps = {
  isLoading: false,
  error: "",
  cards: [],
};

export default Dashboard;
