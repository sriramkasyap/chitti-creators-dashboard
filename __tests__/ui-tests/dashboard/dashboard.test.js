import { render } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";

import Dashboard from "../../../src/components/Dashboard/Dashboard";
import { NewslettersProvider } from "../../../contexts/NewslettersContext";

const cards = [
  {
    id: 1,
    title: "Active Subscribers",
    total: 500,
    icon: "user.png",
  },
  {
    id: 2,
    title: "Paid Subscribers",
    total: 300,
    icon: "user.png",
  },
  {
    id: 3,
    title: "Revenue (per month)",
    total: <span>&#x20b9;1500</span>,
    icon: "money.png",
  },
  {
    id: 4,
    title: "Newsletters Sent",
    total: 29,
    icon: "envelope.png",
  },
];

let container;

beforeEach(() => {
  container = render(
    <NewslettersProvider newsletterStatus={{ status: "draft" }}>
      <Dashboard cards={cards} isLoading={false} error="" />
    </NewslettersProvider>
  );
});

describe("Test Dashboard initial render", () => {
  it("should match the snapshot", () => {
    expect(container).toMatchSnapshot();
  });
});

describe("Test initial getCardDetails api call", () => {
  it("should call getCardDetails", () => {
    expect(container.getAllByTestId("card")).toHaveLength(4);
  });
});

describe("Test props", () => {
  it("should render components on isLoading and error prop change", () => {
    const { rerender, getByTestId } = container;

    rerender(
      <NewslettersProvider newsletterStatus={{ status: "draft" }}>
        <Dashboard cards={cards} isLoading error="" />
      </NewslettersProvider>
    );

    expect(getByTestId("loading_image")).toBeInTheDocument();

    rerender(
      <NewslettersProvider newsletterStatus={{ status: "draft" }}>
        <Dashboard cards={cards} isLoading={false} error="Some Error!" />
      </NewslettersProvider>
    );

    expect(getByTestId("notification")).toBeInTheDocument();
  });
});
