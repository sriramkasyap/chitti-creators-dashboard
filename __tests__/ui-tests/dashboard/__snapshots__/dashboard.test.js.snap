// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Dashboard initial render should match the snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div
      id="chakra-toast-portal"
    >
      <ul
        id="chakra-toast-manager-top"
        style="position: fixed; z-index: 5500; pointer-events: none; display: flex; flex-direction: column; margin: 0px auto;"
      />
      <ul
        id="chakra-toast-manager-top-left"
        style="position: fixed; z-index: 5500; pointer-events: none; display: flex; flex-direction: column;"
      />
      <ul
        id="chakra-toast-manager-top-right"
        style="position: fixed; z-index: 5500; pointer-events: none; display: flex; flex-direction: column;"
      />
      <ul
        id="chakra-toast-manager-bottom-left"
        style="position: fixed; z-index: 5500; pointer-events: none; display: flex; flex-direction: column;"
      />
      <ul
        id="chakra-toast-manager-bottom"
        style="position: fixed; z-index: 5500; pointer-events: none; display: flex; flex-direction: column; margin: 0px auto;"
      />
      <ul
        id="chakra-toast-manager-bottom-right"
        style="position: fixed; z-index: 5500; pointer-events: none; display: flex; flex-direction: column;"
      >
        <li
          class="chakra-toast"
          style="display: flex; flex-direction: column; align-items: flex-end; opacity: 0; transform: translateX(24px) translateZ(0);"
        >
          <div
            class="chakra-toast__inner"
            data-reach-alert="true"
            style="pointer-events: auto; max-width: 560px; min-width: 300px; margin: 0.5rem;"
          >
            <div
              class="css-z9p4wz"
              data-testid="notification"
            >
              fetch is not defined
            </div>
          </div>
        </li>
      </ul>
    </div>
    <div>
      <div
        class="css-i5rilr"
      >
        <div
          class="css-1pnafc6"
        >
          <div
            class="css-1c4iv7r"
            data-testid="card"
          >
            <div
              class="css-z1mrsk"
            >
              <div
                class="css-122cpje"
              >
                <img
                  alt="user.png"
                  class="chakra-image__placeholder css-1hv94z9"
                />
                <p
                  class="chakra-text css-1a5s68p"
                >
                  500
                </p>
              </div>
              <p
                class="chakra-text css-9lx1ak"
              >
                Active Subscribers
              </p>
            </div>
          </div>
          <div
            class="css-1c4iv7r"
            data-testid="card"
          >
            <div
              class="css-z1mrsk"
            >
              <div
                class="css-122cpje"
              >
                <img
                  alt="user.png"
                  class="chakra-image__placeholder css-1hv94z9"
                />
                <p
                  class="chakra-text css-1a5s68p"
                >
                  300
                </p>
              </div>
              <p
                class="chakra-text css-9lx1ak"
              >
                Paid Subscribers
              </p>
            </div>
          </div>
          <div
            class="css-1c4iv7r"
            data-testid="card"
          >
            <div
              class="css-z1mrsk"
            >
              <div
                class="css-122cpje"
              >
                <img
                  alt="money.png"
                  class="chakra-image__placeholder css-1hv94z9"
                />
                <p
                  class="chakra-text css-1a5s68p"
                >
                  <span>
                    ₹1500
                  </span>
                </p>
              </div>
              <p
                class="chakra-text css-9lx1ak"
              >
                Revenue (per month)
              </p>
            </div>
          </div>
          <div
            class="css-1c4iv7r"
            data-testid="card"
          >
            <div
              class="css-z1mrsk"
            >
              <div
                class="css-122cpje"
              >
                <img
                  alt="envelope.png"
                  class="chakra-image__placeholder css-1hv94z9"
                />
                <p
                  class="chakra-text css-1a5s68p"
                >
                  29
                </p>
              </div>
              <p
                class="chakra-text css-9lx1ak"
              >
                Newsletters Sent
              </p>
            </div>
          </div>
        </div>
        <div
          class="css-rfsj4c"
        >
          <div
            class="css-k008qs"
          >
            <h2
              class="chakra-heading css-1pubr08"
            >
              Newsletters List
            </h2>
          </div>
          <div
            class="css-zplcq1"
          >
            <div
              class="css-h53f3m"
            >
              <p
                class="chakra-text css-6q2090"
                data-testid="no-newsletters-data"
              >
                You haven't created any newsletters yet.
                 
                <a
                  class="chakra-link css-1081t4c"
                >
                  Create One now.
                </a>
              </p>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="css-i5rilr"
    >
      <div
        class="css-1pnafc6"
      >
        <div
          class="css-1c4iv7r"
          data-testid="card"
        >
          <div
            class="css-z1mrsk"
          >
            <div
              class="css-122cpje"
            >
              <img
                alt="user.png"
                class="chakra-image__placeholder css-1hv94z9"
              />
              <p
                class="chakra-text css-1a5s68p"
              >
                500
              </p>
            </div>
            <p
              class="chakra-text css-9lx1ak"
            >
              Active Subscribers
            </p>
          </div>
        </div>
        <div
          class="css-1c4iv7r"
          data-testid="card"
        >
          <div
            class="css-z1mrsk"
          >
            <div
              class="css-122cpje"
            >
              <img
                alt="user.png"
                class="chakra-image__placeholder css-1hv94z9"
              />
              <p
                class="chakra-text css-1a5s68p"
              >
                300
              </p>
            </div>
            <p
              class="chakra-text css-9lx1ak"
            >
              Paid Subscribers
            </p>
          </div>
        </div>
        <div
          class="css-1c4iv7r"
          data-testid="card"
        >
          <div
            class="css-z1mrsk"
          >
            <div
              class="css-122cpje"
            >
              <img
                alt="money.png"
                class="chakra-image__placeholder css-1hv94z9"
              />
              <p
                class="chakra-text css-1a5s68p"
              >
                <span>
                  ₹1500
                </span>
              </p>
            </div>
            <p
              class="chakra-text css-9lx1ak"
            >
              Revenue (per month)
            </p>
          </div>
        </div>
        <div
          class="css-1c4iv7r"
          data-testid="card"
        >
          <div
            class="css-z1mrsk"
          >
            <div
              class="css-122cpje"
            >
              <img
                alt="envelope.png"
                class="chakra-image__placeholder css-1hv94z9"
              />
              <p
                class="chakra-text css-1a5s68p"
              >
                29
              </p>
            </div>
            <p
              class="chakra-text css-9lx1ak"
            >
              Newsletters Sent
            </p>
          </div>
        </div>
      </div>
      <div
        class="css-rfsj4c"
      >
        <div
          class="css-k008qs"
        >
          <h2
            class="chakra-heading css-1pubr08"
          >
            Newsletters List
          </h2>
        </div>
        <div
          class="css-zplcq1"
        >
          <div
            class="css-h53f3m"
          >
            <p
              class="chakra-text css-6q2090"
              data-testid="no-newsletters-data"
            >
              You haven't created any newsletters yet.
               
              <a
                class="chakra-link css-1081t4c"
              >
                Create One now.
              </a>
            </p>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
